snapcraft/internal/lxd.py

```diff
@@ -316,18 +316,88 @@ def _ensure_container(self):
+        remote, container_name = self._container_name.split(':')
+        if remote != 'local':
+            return self._remote_mount(destination, source)
+
         devices = self._get_container_status()['devices']
         if destination not in devices:
             check_call([
                 'lxc', 'config', 'device', 'add', self._container_name,
                 destination, 'disk', 'source={}'.format(source),
                 'path={}'.format(destination)])
 
+    def _remote_mount(self, destination, source):
+        # Generate an SSH key and add it to the container's known keys
+        keyfile = 'id_{}'.format(self._container_name)
+        keyfile_path = os.path.join(self._tmp, keyfile)
+        if not os.path.exists(keyfile_path):
+            # Swallow output and allow exceptions to save stdout
+            check_output(['ssh-keygen', '-o', '-N', '', '-f', keyfile_path])
+
+        # Don't use os.path.join because a Windows client would use \
+        ssh_config = '/root/.ssh'
+        ssh_authorized_keys = ssh_config + '/authorized_keys'
+
+        self._container_run(['mkdir', '-p', ssh_config])
+        self._container_run(['chmod', '700', ssh_config])
+        self._container_run(['tee', '-a', ssh_authorized_keys],
+                            stdin=open('{}.pub'.format(keyfile_path), 'r'))
+        self._container_run(['chmod', '600', ssh_authorized_keys])
+
+        ssh_address = self._get_container_address()
+        logger.info('Connecting to {} via SSH'.format(ssh_address))
+        ssh_cmd = ['ssh', '-C', '-F', '/dev/null',
+                   '-o', 'IdentityFile={}'.format(keyfile_path),
+                   '-o', 'StrictHostKeyChecking=no',
+                   '-o', 'UserKnownHostsFile=/dev/null',
+                   '-o', 'User=root',
+                   '-p', '22', ssh_address]
+
+        # One-off command to verify the connection
+        try:
+            check_output(ssh_cmd + ['ls'])
+        except CalledProcessError as e:
+            raise SnapcraftEnvironmentError('Failed to setup SSH') from e
+
+        # Use sshfs in slave mode inside SSH to reverse mount destination
+        self._container_run(['apt-get', 'install', '-y', 'sshfs'])
+        self._container_run(['mkdir', '-p', destination])
+        # Pipes for sshfs and sftp-server to communicate
+        stdin1, stdout1 = os.pipe()
+        stdin2, stdout2 = os.pipe()
+        try:
+            self._host_run(['/usr/lib/sftp-server'],
+                           stdin=stdin1, stdout=stdout2)
+        except CalledProcessError as e:
+            # XXX: This needs to be extended once we support other distros
+            raise SnapcraftEnvironmentError(
+                'sftp-server could not be run.\n'
+                'On Debian, Ubuntu and derivatives, the package '
+                'openssh-sftp-server needs to be installed.'
+                ) from e
+        self._host_run(ssh_cmd + ['sshfs -o slave -o nonempty :{} {}'.format(
```

Comment:
this command could be run with `lxc exec`, right?
If so, and this is the direction to take, `ssh_cmd` could be removed and that check as well, avoiding the need to add routes.

Reply:
No it can't. I guess I should rename the method.